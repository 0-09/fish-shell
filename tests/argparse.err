# No args is an error
argparse: No option specs were provided
# Missing -- is an error
argparse: Missing -- separator
# Flags but no option specs is an error
argparse: No option specs were provided
# Invalid option specs
argparse: Invalid option spec 'h-' at char '-'
argparse: Short flag '+' invalid, must be alphanum or '#'
argparse: Invalid option spec 'h/help:' at char ':'
argparse: Invalid option spec 'h-help::' at char ':'
argparse: Invalid option spec 'h-help=x' at char 'x'
# --max-args and --min-args work
min-max: Expected at least 1 args, got only 0
min-max: Expected at most 3 args, got 4
min-max: Expected at most 1 args, got 2
# Invalid "#-val" spec
argparse: Short flag '#' does not allow modifiers like '='
# Invalid arg in the face of a "#-val" spec
argparse: Unknown option '-x'
Standard input (line 38): 
argparse '#-val' -- abc -x def
^
