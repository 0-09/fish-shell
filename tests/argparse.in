# Test the `argparse` builtin.

##########
# Start by verifying a bunch of error conditions.

echo '# No args is an error' >&2
argparse

echo '# Missing -- is an error' >&2
argparse h/help

echo '# Flags but no option specs is an error' >&2
argparse -s -- hello

echo '# Invalid option specs' >&2
argparse h-
argparse +help
argparse h/help:
argparse h-help::
argparse h-help=x

echo '# --max-args and --min-args work' >&2
begin
    argparse --name min-max --min-args 1 h/help --
    argparse --name min-max --min-args 1 --max-args 3 h/help -- arg1
    argparse --name min-max --min-args 1 --max-args 3 h/help -- arg1 arg2
    argparse --name min-max --min-args 1 --max-args 3 h/help -- --help arg1 arg2 arg3
    argparse --name min-max --min-args 1 --max-args 3 h/help -- arg1 arg2 -h arg3 arg4
    argparse --name min-max --max-args 1 h/help --
    argparse --name min-max --max-args 1 h/help -- arg1
    argparse --name min-max --max-args 1 h/help -- arg1 arg2
end

echo '# Invalid "#-val" spec' >&2
argparse '#-val=' -- abc -x def

echo '# Invalid arg in the face of a "#-val" spec' >&2
argparse '#-val' -- abc -x def

echo '# Defining a short flag more than once' >&2
argparse 's/short' 'x/xray' 's/long' -- -s -x --long

echo '# Defining a long flag more than once' >&2
argparse 's/short' 'x/xray' 'l/short' -- -s -x --long

echo '# Defining an implicit int flag more than once' >&2
argparse '#-val' 'x/xray' 'v#val' -- -s -x --long

echo '# Defining an implicit int flag with modifiers' >&2
argparse 'v#val=' --

##########
# Now verify that validly formed invocations work as expected.

echo '# No args'
argparse h/help --

echo '# One arg and no matching flags'
begin
    argparse h/help -- help
    set -l
    show $argv
end

echo '# Five args with two matching a flag'
begin
    argparse h/help -- help --help me -h 'a lot more'
    set -l
    show $argv
end

echo '# Required, optional, and multiple flags'
begin
    argparse 'h/help' 'a/abc=' 'd/def=?' 'g/ghk=+' -- help --help me --ghk=g1 --abc=ABC --ghk g2 --d -g g3
    set -l
    show $argv
end

echo '# --stop-nonopt works'
begin
    argparse --stop-nonopt 'h/help' 'a/abc=' -- -a A1 -h --abc A2 non-opt 'second non-opt' --help
    set -l
    show $argv
end

echo '# Implicit int flags work'
for v in (set -l -n); set -e $v; end
argparse '#-val' -- abc -123 def
set -l
for v in (set -l -n); set -e $v; end
argparse 'v/verbose' '#-val' 't/token=' -- -123 a1 --token woohoo --234 -v a2 --verbose
set -l
echo '# Should be set to 987'
for v in (set -l -n); set -e $v; end
argparse 'm#max' -- argle -987 bargle
set -l
echo '# Should be set to 765'
for v in (set -l -n); set -e $v; end
argparse 'm#max' -- argle -987 bargle --max 765
set -l
