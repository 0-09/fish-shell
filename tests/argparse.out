# No args
# One arg and no matching flags
argv help
# Five args with two matching a flag
_flag_h 2
_flag_help 2
argv 'help'  'me'  'a lot more'
# Required, optional, and multiple flags
_flag_a ABC
_flag_abc ABC
_flag_d
_flag_def
_flag_g 'g1'  'g2'  'g3'
_flag_ghk 'g1'  'g2'  'g3'
_flag_h 1
_flag_help 1
argv 'help'  'me'
# --stop-nonopt works
_flag_a A2
_flag_abc A2
_flag_h 1
_flag_help 1
argv 'non-opt'  'second non-opt'  '--help'
# Implicit int flags work
_flag_val 123
argv 'abc'  'def'
_flag_t woohoo
_flag_token woohoo
_flag_v 2
_flag_val -234
_flag_verbose 2
argv 'a1'  'a2'
# Should be set to 987
_flag_m 987
_flag_max 987
argv 'argle'  'bargle'
# Should be set to 765
_flag_m 765
_flag_max 765
argv 'argle'  'bargle'
# Check the exit status from argparse validation
_flag_name max
_flag_value 83
expected argparse return status 57
